crates/ruff_linter/src/rules/flake8_bandit/rules/hardcoded_sql_expression.rs

```diff
@@ -13,7 +13,15 @@ use crate::checkers::ast::Checker;
 use super::super::helpers::string_literal;
 
 static SQL_REGEX: Lazy<Regex> = Lazy::new(|| {
-    Regex::new(r"(?i)\b(select\s.+\sfrom\s|delete\s+from\s|(insert|replace)\s.+\svalues\s|update\s.+\sset\s)")
+    // We pass this generated expression strings like:
+    //  "SELECT " + val + " FROM " + table
+    //  f'delete from table where var = {var}'
+    //  "\n    SELECT *\n    FROM table\n    WHERE var = {}\n    ".format(var)
+    //
+    // To avoid false positives, we:
+    //  - Require the SQL to be at the start of the expression, allowing for tokens that are not a part of the string
+    //  - Require whole-word matches for SQL keywords
+    Regex::new(r#"(?i)\A(\"|f\"|\'|f\'|\\|\\n|\s)*\b(select\s.+\sfrom\s|delete\s+from\s|(insert|replace)\s.+\svalues\s|update\s.+\sset\s)"#)
```

Comment:
One possible thing that might help here is to use verbose mode in the regex. That lets you break things apart using insignificant whitespace and even add comments if you want:

```rust
r#"(?ix)
\A
(\"|f\"|\'|f\'|\\|\\n|\s)*
\b
(
  select\s.+\sfrom\s
  |
  delete\s+from\s
  |
  (insert|replace)\s.+\svalues\s
  |
  update\s.+\sset\s
)
"#
```

Or somethinglike that. Dealer's choice about how best to break it up.

Now that I've written out, it makes me wonder about the following:

* Is it intentional to have the escapes in `\"|f\"|\'|f\'|\\|\\n|\s` here?
Namely, since this is a raw string, `f\"` will match `f\"` literally and not
`f"`. Similarly for the other branches.
* While the regex didn't previously have it, is it perhaps desirable to add a
word boundary assertion at the end as well? Although maybe not, since I see
that each of the `select`, `delete`, `insert|replace` and `update` branches
all must end with a `\s`.

Reply:
Do you have suggestions for handling the false negatives? It looks like the regex would need to get much more complicated. This change may not be worth it as-is.

Reply:
Derp, yes, `f\"` is equivalent to `f"`. (Older versions of the regex crate would actually reject such escapes as superfluous.)

With respect to FNs and FPs, are you constrained to a single regex? If not, you could combine the status quo with what you have here. You could change the status quo to only match for uppercase things like `SELECT`. That will reduce its false positives but greatly increase the false negatives. Then to decrease false negatives, you could use your second regex here which allows lowercase `select` but only at the start of a string.

It's pretty hokey and probably is an overfitting to our current tests but is perhaps an improvement.

Reply:
Hmm... definitely hokey but I was also thinking case sensitivity makes some sense...

I almost would rather have a "case insensitive" setting for this which is off by default...

Reply:
For f-strings, we wouldn't need to worry about the prefix and quotes as in https://github.com/astral-sh/ruff/pull/7927/files#diff-649016c25b517ba99e70caf422c95ef9d544e7c806d26e64624c0414777d54f5 I've removed them. That is, for f-strings, we'd do:

```rs
/// Concatenates the contents of an f-string, without the prefix and quotes,
/// and escapes any special characters.
///
/// ## Example
///
/// ```python
/// "foo" f"bar {x}" "baz"
/// ```
///
/// becomes `foobar {x}baz`.
```